### DISCUSSION TITLE
Move the window when no default title bar is in place

### DISCUSSION THREAD

**Discussion by GMas0124816** [AvaloniaUI/Avalonia#8441] [Q&A] [answered]
Created a custom title bar and all works just fine, but I would like the user to have the option to move the window when the pointer is pressed on 'my title bar' and dragged around. I found that I could go a long way with the point position of the 'OnPointerPressed' and 'OnPointerReleased' eventargs values, but I do not find the variable of the window to change the x, y value of the window. Position is read only.
Any suggestions ?

**Comment by maxkatz6** [comment]
On some platforms you can use `Window.BeginMoveDrag(pointerEventArgs)`. Works on Windows, but it might fail on some linux distributives and won't work on macOS.

> Position is read only.

Is it? https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Window.cs#L443-L446

**Comment by GMas0124816** [comment, marked as answer]
@maxkatz6 The position is indeed not read only, only the X and Y values, so one needs to create a new PixelPoint and assign it to the position of the form. My mistake !

This code works on everything I assume as it only uses high level events. Just put it in the windows cs file and link the according events in your xaml code.

```
private bool _mouseDownForWindowMoving = false;
private PointerPoint _originalPoint;

private void InputElement_OnPointerMoved(object? sender, PointerEventArgs e)
 {
            if (!_mouseDownForWindowMoving) return;

            PointerPoint currentPoint = e.GetCurrentPoint(this);
            Position = new PixelPoint(Position.X + (int)(currentPoint.Position.X - _originalPoint.Position.X),
                Position.Y + (int)(currentPoint.Position.Y - _originalPoint.Position.Y));
}

private void InputElement_OnPointerPressed(object? sender, PointerPressedEventArgs e)
{
            if (WindowState == WindowState.Maximized || WindowState == WindowState.FullScreen) return;

            _mouseDownForWindowMoving = true;
            _originalPoint = e.GetCurrentPoint(this);
}

private void InputElement_OnPointerReleased(object? sender, PointerReleasedEventArgs e)
{
            _mouseDownForWindowMoving = false;
}
```

OnPointerPressed needs to have the condition of the windows state since in case of multi monitor, if you move the window (in maximized or full screen state) from NOT the main monitor to another one, the window turns white (and remains white), but only in Maximized or Full screen state. So moving it around in 'Normal' state now just works fine on all monitors and the starting of the dragging now can be from any monitor.

**Reply by th3w1zard1** [reply]
This is a fantastic answer, the only problem is if a user clicks on a control (e.g. a combobox) the window move code will start, making it impossible to select an option.

**Reply by timunie** [reply]
you can check `sender` for it's type. If it's any input control, just skip the dragging logic. if `sender` is not enough, check also `e.OriginalSource`

**Reply by pvogt** [reply]
For those having issues using `Window.BeginMoveDrag(pointerEventArgs)` and trying to return a Pointer Released Event of a control, the Accepted Answer is an excellent alternative as it allows for moving of your App Window and triggering the Pointer Released Event when finished.

**Reply by NathanielHardesty** [reply]
Just as a note for others who find this, the exact behavior of `Window.BeginMoveDrag(pointerEventArgs)` on Microsoft Windows, where it will unmaximize the window, position it, disable the usual unmaximize animation etc, is extremely difficult to replicate. Your best option may be to check for the operating system using `OperatingSystem.IsWindows()` and go from there.

**Reply by NeverMorewd** [reply]
> Window.BeginMoveDrag was obsolute

**Reply by JosiahDanger** [reply]
Thank you @GMas0124816 for your accepted answer. I'm leaving a comment here because I have identified two means of simplifying your code. I hope that these suggestions prove helpful to anyone who may discover my comment.

1. In the accepted answer, it is redundant to create variables of type `PointerPoint`, because the developer needs only to access their Position properties. Instead, the developer may call `e.GetPosition(this)`, which returns a `Point`.
2. It is possible to convert a `Point` to a `PixelPoint` by calling on the window in question `this.PointToScreen(yourPointHere)`.

I am going to copy-paste my suggested changes. Feel free to stick this in your '.cs' file.

```
    private bool _isWindowDragInEffect = false;
    private Point _cursorPositionAtWindowDragStart = new(0, 0);

    private void WindowDragHandle_OnPointerMoved(object? sender, PointerEventArgs e)
    {
        if (_isWindowDragInEffect)
        {
            Point currentCursorPosition = e.GetPosition(this);
            Point cursorPositionDelta = currentCursorPosition - _cursorPositionAtWindowDragStart;

            Position = this.PointToScreen(cursorPositionDelta);
        }
    }

    private void WindowDragHandle_OnPointerPressed(object? sender, PointerPressedEventArgs e)
    {
        if ((e.Source is Control sourceControl) && (sourceControl.Name == "WindowDragHandle"))
        {
            _isWindowDragInEffect = true;
            _cursorPositionAtWindowDragStart = e.GetPosition(this);
        }
    }

    private void WindowDragHandle_OnPointerReleased(object? sender, PointerReleasedEventArgs e) =>
        _isWindowDragInEffect = false;
```
